### PR TITLE
388 consent portal add consentcontext react context for consent management

### LINKED ISSUE #388: [Consent Portal] Add ConsentContext (React Context) for Consent Management
### Current Limitation

Currently, the state handling and consent handling is tied to App.tsx.

### Suggested Improvement

Need to refactor that part to a React Context such that it centralizes the complete consent handling mechanism.

### Version

v0.0.1

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mushrafmim**
## Summary
Needs a centralized management of consent, (Extracted consent management to a React Context API)

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Other (please describe):

## Changes Made
- Add React Router Dom for Routing
- Extract ConsentRecord to `types.ts` file
- Create `ConsentContext` to manage consent.

## Testing
- [x] I have tested this change locally
- [x] I have added unit tests for new functionality
- [x] I have tested edge cases
- [x] All existing tests pass

## Checklist
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have checked that there are no merge conflicts

## Related Issues
Closes #388

### CODE REVIEW

File: portals/consent-portal/src/ConsentContext.tsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    try {
+      const headers = await getAuthHeaders();
+      const response = await fetch(`${CONSENT_ENGINE_PATH}/consents/${consentUuid}`, {headers});
+
+      if (!response.ok) {
+        let errorMessage = '';
+        try {
+          const errorData = await response.json();
+          errorMessage = errorData.error || '';
+        } catch { /* empty */
```
Comment (ginaxu1) [inline portals/consent-portal/src/ConsentContext.tsx]: why error handling in `fetchConsentData` and `handleConsentDecision` not properly handle all error cases? some errors may be silently ignored.

suggest

```typescript
let errorMessage = '';
try {
  const errorData = await response.json();
  errorMessage = errorData.error || errorData.message || `HTTP ${response.status}`;
} catch (parseError) {
  // If JSON parsing fails, use status text
  errorMessage = response.statusText || `HTTP ${response.status}`;
  console.warn('Failed to parse error response:', parseError);
}
```

File: portals/consent-portal/src/ConsentContext.tsx [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  const navigate = useNavigate();
+  const [consentRecord, setConsentRecord] = useState<ConsentRecord | null>(null);
+  const [error, setError] = useState('');
+  const [consentId, setConsentId] = useState<string | null>(null);
+  const [userEmail, setUserEmail] = useState<string>('');
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const {isSignedIn, user, getAccessToken} = useAsgardeo();
+
+  const CONSENT_ENGINE_PATH = window?.configs?.apiUrl;
```
Comment (ginaxu1) [inline portals/consent-portal/src/ConsentContext.tsx]: config is accessed via `window?.configs?.apiUrl` without validation or fallback. suggest:

```typescript
const getConsentEnginePath = (): string => {
  const apiUrl = window?.configs?.apiUrl;
  if (!apiUrl) {
    throw new Error('CONSENT_ENGINE_PATH is not configured');
  }
  try {
    new URL(apiUrl); // Validate URL format
    return apiUrl;
  } catch {
    throw new Error(`Invalid CONSENT_ENGINE_PATH: ${apiUrl}`);
  }
};

const CONSENT_ENGINE_PATH = getConsentEnginePath();
```

File: portals/consent-portal/src/ConsentContext.tsx [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    if (!consentUuid) {
+      setError('No consent ID provided.');
+      navigate('/error');
+      return;
+    }
+    return fetchConsentData(consentUuid);
+  };
+
+  const handleConsentDecision = async (decision: 'approved' | 'rejected') => {
```
Comment (ginaxu1) [inline portals/consent-portal/src/ConsentContext.tsx]: `handleConsentDecision` doesn't validate the payload before sending to the server, suggest:

```typescript
const handleConsentDecision = async (decision: 'approved' | 'rejected') => {
  if (!consentRecord) {
    setError('Missing consent data');
    return;
  }
  
  if (!consentRecord.consent_id) {
    setError('Invalid consent record: missing consent_id');
    return;
  }
  
  if (!userEmail || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(userEmail)) {
    setError('Invalid user email');
    return;
  }
  
  // Send only required fields
  const payload = {
    action: decision,
    updated_by: userEmail,
    reason: decision === 'rejected'
      ? 'User denied the consent request via self-service portal'
      : 'User approved the consent request via self-service portal'
  };
  // ...
};
```

File: portals/consent-portal/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  status: 'pending' | 'approved' | 'rejected' | 'expired' | 'revoked';
+  type?: string;
+  created_at: string;
+  updated_at: string;
+  expires_at: string;
+  fields: string[];
+  session_id: string;
+  redirect_url: string;
+  app_display_name: string;
+}
```
Comment (ginaxu1) [inline portals/consent-portal/src/types.ts]: nit: end this file with newline. run `cd portals/consent-portal && npm run lint -- --fix` to auto fix formatting issues

File: portals/consent-portal/src/ConsentContext.tsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    }
+  };
+
+  const updateConsentId = (consentId: string) => {
+    setConsentId(consentId);
+    // Store the consent_id in localStorage for persistence across redirects
+    localStorage.setItem('consent_id', consentId);
+  };
+
+  useEffect(() => {
```
Comment (ginaxu1) [inline portals/consent-portal/src/ConsentContext.tsx]: same as above useEffect, suggest

```
useEffect(() => {
  let isMounted = true;
  
  if (isSignedIn) {
    fetchUserInfo();
  }
  
  return () => {
    isMounted = false;
  };
}, [fetchUserInfo, isSignedIn]);
```

File: portals/consent-portal/src/ConsentContext.tsx [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    return headers;
+  }, [getAccessToken]);
+
+  const fetchUserInfo = useCallback(() => {
+    if (user) {
+      setUserEmail(user.email || '');
+    }
+  }, [user]);
+
+  const fetchConsentData = useCallback(async (consentUuid: string) => {
```
Comment (ginaxu1) [inline portals/consent-portal/src/ConsentContext.tsx]: use `AbortController` for async operations in `fetchConsentData`, suggest

```typescript
const fetchConsentData = useCallback(async (consentUuid: string, signal?: AbortSignal) => {
  try {
    const headers = await getAuthHeaders();
    const response = await fetch(`${CONSENT_ENGINE_PATH}/consents/${consentUuid}`, {
      headers,
      signal
    });
    // ... rest of code
  } catch (err) {
    if (err instanceof Error && err.name === 'AbortError') {
      return null; // Request was cancelled
    }
    // ... error handling
  }
}, [CONSENT_ENGINE_PATH, getAuthHeaders, navigate]);
```

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

File: portals/consent-portal/src/ConsentContext.tsx (hunk 2 above)
Comment (mushrafmim) [reply portals/consent-portal/src/ConsentContext.tsx]: This is handled in useEffect hook.
